### PR TITLE
Add CachingCostBalancerStrategy

### PR DESCRIPTION
**Pull request by dgolitsyn**
CachingCostBalancerStrategy is fast implementation of CostBalancerStrategy. Computation algorithm slightly differs from the original version and allows pre-compute and cache cost function values to reduce complexity of the algorithm.

Motivation:
 1. Current implementation is very complex computational task and calculation time grows linearly with number of segments in the cluster. For rather big cluster one druid balancer iteration could take tens of minutes
 2. If large number of segments need to be rebalanced, cost function calculation could take hours or even days (if number of historical nodes changed for instance)
 3. Total server cost could not be calculated due to high computational complexity and thus could not be included in balancing algorithm

Implementation details:
 - Result numbers are not equal for CachingCostBalancerStrategy and CostBalancerStrategy, but decisions made are the same, meaning that the same server has been chosen for particular segment by two strategies (has been tested for ~100k segments from our cluster)
 - CachingCostBalancerStrategy is less memory efficient and requires ~100mb heap for each million of segments
 - Computation time is ~10000x faster (depends on segment distribution and total number of servers and segments)

Benchmark

```
cachingCostStrategy  avgt   10      1.766 ±    0.035  us/op
costStrategy         avgt   10  21987.245 ± 3593.512  us/op
```

### CODE REVIEW

File: server/src/main/java/io/druid/server/coordinator/cost/SegmentsCostCache.java [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+      index = (index >= 0) ? index : -index - 1;
+      return addLeftCost(dataSegment, t0, t1, index) + rightCost(dataSegment, t0, t1, index);
+    }
+
+    private double addLeftCost(DataSegment dataSegment, double t0, double t1, int index)
+    {
+      double leftCost = 0.0;
+      // add to cost all left-overlapping segments
+      int leftIndex = index - 1;
+      while (leftIndex >= 0
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/coordinator/cost/SegmentsCostCache.java]: This loop looks pretty odd, can it simply be

```java
for (int leftIndex = index - 1; leftIndex >=0; --leftIndex) {
final Segment segment = sortedSegments.get(leftIndex);
if (!segment.getInterval().overlaps(dataSegment.getInterval()))
{
  break;
}
final double start = convertStart(sortedSegments.get(leftIndex), interval);
final double end = convertEnd(sortedSegments.get(leftIndex), interval);
leftCost += CostBalancerStrategy.intervalCost(end - start, t0 - start, t1 - start);
}
```
or similar?

File: server/src/main/java/io/druid/server/coordinator/cost/SegmentsCostCache.java [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+      }
+      return leftCost;
+    }
+
+    private double rightCost(DataSegment dataSegment, double t0, double t1, int index)
+    {
+      double rightCost = 0.0;
+      // add all right-overlapping segments
+      int rightIndex = index;
+      while (rightIndex < sortedSegments.size() &&
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/coordinator/cost/SegmentsCostCache.java]: same thing about using a `for` loop for readability.

File: server/src/main/java/io/druid/server/coordinator/cost/ClusterCostCache.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    Builder builder = builder();
+    segmentsByServerName.forEach(
+        (serverName, segments) -> segments.forEach(segment -> builder.addSegment(serverName, segment))
+    );
+    return builder;
+  }
+
+  public static class Builder
+  {
+    private final Map<String, ServerCostCache.Builder> serversCostCache = new HashMap<>();
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/coordinator/cost/ClusterCostCache.java]: How is this guarded when https://github.com/druid-io/druid/pull/4731/files#diff-9ef443ad87a0cafc15b5318fb10590c5R122 has multiple threads?

File: server/src/main/java/io/druid/server/coordinator/CachingCostBalancerStrategyFactory.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    }
+  }
+
+  @LifecycleStop
+  public void stop()
+  {
+    if (!lifecycleLock.canStop()) {
+      throw new ISE("CachingCostBalancerStrategyFactory can not be stopped");
+    }
+    executor.shutdownNow();
```
[the remark this answers is not in the document]
Comment (leventov) [reply server/src/main/java/io/druid/server/coordinator/CachingCostBalancerStrategyFactory.java]: exitStop() should be called only if lifecycled object is recycleable

File: server/src/main/java/io/druid/server/coordinator/cost/SegmentsCostCache.java (hunk 1 above)
Comment (leventov) [reply server/src/main/java/io/druid/server/coordinator/cost/SegmentsCostCache.java]: It uses `leftCost` variable outside of the loop

File: server/src/main/java/io/druid/server/coordinator/CachingCostBalancerStrategyFactory.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (leventov) [reply server/src/main/java/io/druid/server/coordinator/CachingCostBalancerStrategyFactory.java]: Added a comment

File: server/src/main/java/io/druid/client/AbstractCuratorServerInventoryView.java [hunk 5]
```diff
@@ -249,7 +249,7 @@ public void run()
 
   protected void runServerCallbacks(final DruidServer server)
```
[the remark this answers is not in the document]
Comment (leventov) [reply server/src/main/java/io/druid/client/AbstractCuratorServerInventoryView.java]: Renamed

File: server/src/main/java/io/druid/server/coordinator/cost/SegmentsCostCache.java [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+   */
+  private static final double HALF_LIFE_DAYS = 1.0;
+  private static final double LAMBDA = Math.log(2) / HALF_LIFE_DAYS;
+  private static final double MILLIS_FACTOR = TimeUnit.DAYS.toMillis(1) / LAMBDA;
+
+  /**
+   * LIFE_THRESHOLD is used to avoid calculations for segments that are "far"
+   * from each other and thus cost(X,Y) ~ 0 for these segments
+   */
+  private static final long LIFE_THRESHOLD = TimeUnit.DAYS.toMillis(30);
```
[the remark this answers is not in the document]
Comment (leventov) [reply server/src/main/java/io/druid/server/coordinator/cost/SegmentsCostCache.java]: Agreed

File: server/src/main/java/io/druid/server/coordinator/cost/SegmentsCostCache.java [hunk 7]
```diff
@@ -0,0 +1,441 @@
...
+      builder.addSegment(segment);
+      return this;
+    }
+
+    public Builder removeSegment(DataSegment segment)
+    {
+      Interval interval = getBucketInterval(segment);
+      buckets.computeIfPresent(
+          interval,
+          (i, builder) -> builder.removeSegment(segment).isEmpty() ? null : builder
```
[the remark this answers is not in the document]
Comment (leventov) [reply server/src/main/java/io/druid/server/coordinator/cost/SegmentsCostCache.java]: Remove the key, added a comment

File: server/src/main/java/io/druid/server/coordinator/cost/ClusterCostCache.java (hunk 3 above)
Comment (leventov) [reply server/src/main/java/io/druid/server/coordinator/cost/ClusterCostCache.java]: This exec is not used to populate the builder. Another one, stored in a field in `CachingCostBalancerStrategyFactory` is always single-threaded, added a comment about that.

**Comment by leventov** [comment]
@drcrallen thanks for review, addressed comments

File: server/src/main/java/io/druid/server/coordinator/cost/SegmentsCostCache.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (drcrallen) [reply server/src/main/java/io/druid/server/coordinator/cost/SegmentsCostCache.java]: Is that guaranteed anywhere, or just specific to some implementations? Here's where my java8 knowledge falls flat.

[the remark this answers is not in the document]
Comment (leventov) [reply server/src/main/java/io/druid/server/coordinator/cost/SegmentsCostCache.java]: Yes, it's guaranteed by `java.util.Map` spec: https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#computeIfPresent-K-java.util.function.BiFunction-

[the remark this answers is not in the document]
Comment (drcrallen) [reply server/src/main/java/io/druid/server/coordinator/cost/SegmentsCostCache.java]: Ah yes, I missed it on my other read through. thanks!

**Comment by drcrallen** [comment]
If this is the same as the prior cost balancer strategy, can it reuse the tests from `io.druid.server.coordinator.CostBalancerStrategyTest`?

**Comment by leventov** [comment]
It doesn't produce the same numbers, but (supposedly) makes the same decisions. That's what `CachingCostBalancerStrategyTest` verifies.

**Pull request merged**